### PR TITLE
CAP-360: Fix Aircraft spatial extent

### PR DESCRIPTION
**Pull request by tomw66**
Fixes [this issue](https://metoffice.atlassian.net/browse/CAP-360?atlOrigin=eyJpIjoiOTRjYWVkMDVkNmVmNDE2YmEwOGE2YWE3ZTk5ZGZkNDUiLCJwIjoiaiJ9) when extracting aircraft metadata.

Was being caused by some of the latitude & longitude values not being numbers (looks like they are just underscores, presumably when the aircraft can’t get an accurate location reading). This then meant we were getting an error when trying to calculate the spatial extent (as in the area covered) of the flight, as the code was expecting to only have to deal with numbers, and would fail if lon or lat was a masked array with a NaN at the start or end.

[numpy.ma.min/max](https://numpy.org/doc/stable/reference/generated/numpy.ma.min.html) should be able to find the min & max values of either a standard array, or masked array with NaNs; indeed, all aircraft files that can be loaded with iris can now be extracted!

I also realised that the code was wrong, even when passing - we were using the first and final coordinates of the data to find spatial extent, rather than the min and max - this is especially useless for flights, which are likely taking off and landing in the same place!

For an accompanying test I've generalised it a bit to test on a subset of actual flight data; this should be a good way to ensure `extract_metadata` remains healthy overall, so planning on doing this for every data type/source.

### CODE REVIEW

**Comment by tomw66** [comment]
@corinnebosley I think you know more about masked arrays than me so might be able to comment if this is a good idea! Typical dimension data when missing values is like:
`AuxCoord(masked_array(data=[50.78006, 50.78006, 50.78006, ..., --, --, --],
             mask=[False, False, False, ...,  True,  True,  True],
       fill_value=nan), standard_name='latitude', units=Unit('degrees'), long_name='aircraft_latitude_measured_by_the_AIMMS', var_name='AIMMS_latitude', attributes={'frequency': '1 Hz'})`

**Comment by tomw66** [comment]
@corinnebosley actually I'm thinking maybe it's best to move this aircraft nc file to cap-sample-data and import from there. Can you show me how to do this when youre next around?

File: src/clean_air/data/extract_metadata.py [hunk 1]
```diff
@@ -106,11 +109,21 @@ def extract_metadata(
...
+        # list of coords that may represent z axis, in ascending order of preference
+        z_options = [cube.coords('altitude'), cube.coords(axis='z')]
+        for item in z_options:
+            if len(item) == 1:
+                z_coord = item[0].points
+        if z_coord is not None:
+            if np.ma.isMaskedArray(z_coord):
+                z_coord = z_coord.compressed()
+            vertical_extent = VerticalExtent(z_coord)
+            total_vertical_extent_set.update(z_coord)
```
Comment (corinnebosley) [inline src/clean_air/data/extract_metadata.py:126]: Have to admit, I don't fully understand some of this (not familiar with `coord.compressed`, and I don't quite understand the looping here) but I trust your judgement.

**Review by corinnebosley — CHANGES_REQUESTED** [review]
Work looks pretty good to me.  Nice fix.

Currently the test is failing, but I assume (and have had a look to check) that this is just because the aircraft file needs to be added to cap-sample-data.

To do this, you need to clone the cap-sample-data repo if you haven't already done so, then just make a new branch, chuck the file in the 'aircraft' directory and then commit and push.  You will also need to make sure that the test is now pulling the file from cap-sample-data, but let me know if you have any problems along the way.

File: src/clean_air/data/extract_metadata.py (hunk 1 above)
Comment (tomw66) [reply src/clean_air/data/extract_metadata.py:126]: This part just checks if our z coordinate data is a masked array, and if it is, uses this [function](https://numpy.org/doc/stable/reference/generated/numpy.ma.MaskedArray.compressed.html) to get a ndarray of our data that will then work with our `VerticalExtent` object!

Comment (corinnebosley) [reply src/clean_air/data/extract_metadata.py:126]: Cool.  I like that!

**Review by corinnebosley — APPROVED** [review]

**Comment by tomw66** [comment]
In the cap_env yaml, I've constrained shapely to <v2, which will fix the 2 non-CRS [failing tests](https://shapely.readthedocs.io/en/stable/migration.html). However, this environment change is not carried through to the github actions run, where tox is still using shapely v2, so they're still failing. How fun!

I'm completely stumped by the remaining failures. I've pinned all of the relevant packages I can think of to their versions when they were passing:

- numpy==1.23.5
- shapely==1.8.5
- pyproj==3.4.0
- geos==3.11.1

But these have no impact, so haven't pushed changes.

File: environment.yml [hunk 2]
```diff
@@ -8,10 +8,12 @@ dependencies:
 
 
   # Main data handling
+  - pyproj==3.4.0
   - iris
   - xarray
   - pandas
   - geopandas
+  - "shapely<2"
```
Comment (corinnebosley) [inline environment.yml:16]: Does this need to be in quote marks?  Not sure what difference it makes, but it might.

**Comment by corinnebosley** [comment]
@tomw66 I have felt the pain of fixing environments, so when you're ready today we can work on this one together if you like, see if we can figure out what's going on.  Your call.

**Pull request merged**